loki/transformations/tests/test_pool_allocator.py

```diff
@@ -348,6 +349,112 @@ def test_pool_allocator_temporaries(tmp_path, frontend, generate_driver_stack, b
         assert 'stop' not in fcode.lower()
 
 
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_pool_allocator_unused_temporaries(tmp_path, frontend, horizontal, block_dim):
+    fcode_driver = """
+subroutine driver(NLON, NZ, NB, FIELD1, FIELD2)
+    use kernel_mod, only: kernel
+    implicit none
+    INTEGER, PARAMETER :: JPRB = SELECTED_REAL_KIND(13,300)
+    INTEGER, INTENT(IN) :: NLON, NZ, NB
+    real(kind=jprb), intent(inout) :: field1(nlon, nb)
+    real(kind=jprb), intent(inout) :: field2(nlon, nz, nb)
+    integer :: b
+    do b=1,nb
+        call KERNEL(1, nlon, nlon, nz, field1(:,b), field2(:,:,b))
+    end do
+end subroutine driver
+    """.strip()
+    fcode_kernel = """
+module kernel_mod
+    implicit none
+contains
+    subroutine kernel(start, end, klon, klev, field1, field2)
+        implicit none
+        integer, parameter :: jprb = selected_real_kind(13,300)
+        integer, parameter :: nclv = 2
+        integer, intent(in) :: start, end, klon, klev
+        real(kind=jprb), intent(inout) :: field1(klon)
+        real(kind=jprb), intent(inout) :: field2(klon,klev)
+        real(kind=jprb) :: tmp1(klon)
+        real(kind=jprb) :: tmp2(klon, klev)
+        ! shouldn't end up on stack since not used at all
+        real(kind=jprb) :: tmp3(klon, klev)
+        ! shouldn't end up on stack since all dimension are constant
+        real(kind=jprb) :: tmp4(nclv)
+        ! shouldn't end up on stack since all dimension are constant
+        real(kind=jprb) :: tmp5(2)
+        ! should be on the stack although only read and not written to
+        real(kind=jprb) :: tmp6(klon, nclv)
+        ! should be on the stack although only used as an argument in a call
+        real(kind=jprb) :: tmp7(klon, nclv)
```

Comment:
This is very nice, covering the matrix of cases with comments explaining what's going on and to expect! ❤️

Reply:
⬆️ What he said! Very neat! 🙏